Sources/Signal.swift

```diff
@@ -2015,177 +1876,346 @@ private enum ThrottleWhileState<Value> {
 	}
 }
 
+private protocol SignalAggregateStrategy {
+	/// Update the latest value of the signal at `position` to be `value`.
+	///
+	/// - parameters:
+	///   - value: The latest value emitted by the signal at `position`.
+	///   - position: The position of the signal.
+	///
+	/// - returns: `true` if the aggregating signal should terminate as a result of the
+	///            update. `false` otherwise.
+	mutating func update(_ value: Any, at position: Int) -> Bool
+
+	/// Record the completion of the signal at `position`.
+	///
+	/// - parameters:
+	///   - position: The position of the signal.
+	///
+	/// - returns: `true` if the aggregating signal should terminate as a result of the
+	///            completion. `false` otherwise.
+	mutating func complete(at position: Int) -> Bool
+
+	init(count: Int, action: @escaping (ContiguousArray<Any>) -> Void)
+}
+
 extension Signal {
+	private struct CombineLatestStrategy: SignalAggregateStrategy {
+		private enum Placeholder {
+			case none
+		}
+
+		private var values: ContiguousArray<Any>
+		private var completionCount: Int
+		private let action: (ContiguousArray<Any>) -> Void
+
+		private var _hasAllSentInitial: Bool
+		private var hasAllSentInitial: Bool {
+			mutating get {
+				if _hasAllSentInitial {
+					return true
+				}
+
+				_hasAllSentInitial = values.reduce(true) { $0 && !($1 is Placeholder) }
+				return _hasAllSentInitial
+			}
+		}
+
+		mutating func update(_ value: Any, at position: Int) -> Bool {
+			values[position] = value
+
+			if hasAllSentInitial {
+				action(values)
+			}
+
+			return false
+		}
+
+		mutating func complete(at position: Int) -> Bool {
+			completionCount += 1
+			return completionCount == values.count
+		}
+
+		init(count: Int, action: @escaping (ContiguousArray<Any>) -> Void) {
+			values = ContiguousArray(repeating: Placeholder.none, count: count)
+			completionCount = 0
+			_hasAllSentInitial = false
+			self.action = action
+		}
+	}
+
+	private struct ZipStrategy: SignalAggregateStrategy {
+		private var values: ContiguousArray<[Any]>
+		private var isCompleted: ContiguousArray<Bool>
+		private let action: (ContiguousArray<Any>) -> Void
+
+		private var hasCompletedAndEmptiedSignal: Bool { return Swift.zip(values, isCompleted).contains(where: { $0.isEmpty && $1 }) }
```

Comment:
Would you mind putting these over multiple lines? I think that makes them more readable.